File: frontend/types/strategy-research.ts
export type ResearchStrategy = {
  id: number;
  name: string;
  template: string;
  status: "CANDIDATE" | "PROMOTED" | "REJECTED" | "ARCHIVED" | string;
  origin: string;
  family_id: number | null;
  best_fitness: string | number;
  best_version_id: number | null;
  parameters: Record<string, number>;
  created_at: string;
  updated_at: string;
};

export type StrategyVersion = {
  id: number;
  strategy_id: number;
  version: number;
  parameters: Record<string, number>;
  sharpe: string | number;
  profit_factor: string | number;
  max_drawdown: string | number;
  stability_score: string | number;
  consistency_score: string | number;
  fitness: string | number;
  overfit: boolean;
  total_trades: number;
  created_at: string;
};

export type ResearchExperiment = {
  id: number;
  experiment_type: string;
  status: string;
  strategy_id: number | null;
  symbol: string;
  timeframe: string;
  result: Record<string, unknown>;
  fitness: string | number;
  created_at: string;
  completed_at: string | null;
};

export type FeatureRecord = {
  id: number;
  name: string;
  category: string;
  symbol: string;
  timeframe: string;
  importance_score: string | number;
  correlation_with_profit: string | number;
  stability_score: string | number;
  updated_at: string;
};

export type HypothesisTest = {
  id: number;
  statement: string;
  feature: string;
  condition: string;
  status: "SUPPORTED" | "REJECTED" | "INCONCLUSIVE" | "UNTESTED" | string;
  supported: boolean;
  edge: string | number;
  p_value: string | number;
  sample_size: number;
  result: Record<string, unknown>;
  created_at: string;
};

export type ResearchRunResult = {
  evaluated: number;
  promoted: number;
  best_fitness: number;
  best_strategy_id: number | null;
  best_sharpe: number | null;
  reason: string | null;
  leaderboard: Array<{
    strategy_id: number;
    name: string;
    fitness: number;
    sharpe: number;
    overfit: boolean;
  }>;
};

export type PromotionResult = {
  strategy_id: number;
  decision: string;
  passed: boolean;
  reasons: string[];
};

File: backend/app/schemas/strategy_research.py
from datetime import datetime
from decimal import Decimal

from pydantic import BaseModel


class GenerateIn(BaseModel):
    template: str = "ema_rsi_atr"
    symbol: str = "BTC_USDT"
    timeframe: str = "1h"
    feature_driven: bool = False
    evaluate: bool = True


class RunIn(BaseModel):
    symbol: str = "BTC_USDT"
    timeframe: str = "1h"
    population: int | None = None


class RunOut(BaseModel):
    evaluated: int
    promoted: int
    best_fitness: float
    best_strategy_id: int | None = None
    best_sharpe: float | None = None
    reason: str | None = None
    leaderboard: list[dict]


class StrategyVersionOut(BaseModel):
    id: int
    strategy_id: int
    version: int
    parameters: dict
    sharpe: Decimal
    profit_factor: Decimal
    max_drawdown: Decimal
    stability_score: Decimal
    consistency_score: Decimal
    fitness: Decimal
    overfit: bool
    total_trades: int
    created_at: datetime

    class Config:
        from_attributes = True


class ResearchStrategyOut(BaseModel):
    id: int
    name: str
    template: str
    status: str
    origin: str
    family_id: int | None
    best_fitness: Decimal
    best_version_id: int | None
    parameters: dict
    created_at: datetime
    updated_at: datetime

    class Config:
        from_attributes = True


class ExperimentOut(BaseModel):
    id: int
    experiment_type: str
    status: str
    strategy_id: int | None
    symbol: str
    timeframe: str
    result: dict
    fitness: Decimal
    created_at: datetime
    completed_at: datetime | None

    class Config:
        from_attributes = True


class FeatureRecordOut(BaseModel):
    id: int
    name: str
    category: str
    symbol: str
    timeframe: str
    importance_score: Decimal
    correlation_with_profit: Decimal
    stability_score: Decimal
    updated_at: datetime

    class Config:
        from_attributes = True


class HypothesisTestOut(BaseModel):
    id: int
    statement: str
    feature: str
    condition: str
    status: str
    supported: bool
    edge: Decimal
    p_value: Decimal
    sample_size: int
    result: dict
    symbol: str
    timeframe: str
    created_at: datetime

    class Config:
        from_attributes = True


class ABTestOut(BaseModel):
    id: int
    strategy_a_id: int | None
    strategy_b_id: int | None
    symbol: str
    timeframe: str
    winner: str
    a_fitness: Decimal
    b_fitness: Decimal
    p_value: Decimal
    detail: str
    created_at: datetime

    class Config:
        from_attributes = True


class PromotionOut(BaseModel):
    strategy_id: int
    decision: str
    passed: bool
    reasons: list[str]


class StrategyDetailOut(BaseModel):
    strategy: ResearchStrategyOut
    versions: list[StrategyVersionOut]
    trades: list[dict]
    equity_curve: list[dict]

    class Config:
        from_attributes = True


class CustomHypothesisIn(BaseModel):
    statement: str
    feature: str
    condition_desc: str
    expects_negative: bool = False
    symbol: str = "BTC_USDT"
    timeframe: str = "1h"


class SymbolOut(BaseModel):
    symbol: str
    has_data: bool
